packages/fluentui/react-northstar/test/specs/commonTests/utils/getComponentPath.tsx

```diff
@@ -0,0 +1,68 @@
+import * as React from 'react';
+import { mount } from 'enzyme';
+import config from '@uifabric/build/config';
+
+// hack to work around lack of esModuleInterop
+import * as StackUtilsType from 'stack-utils'; // eslint-disable-line
+const StackUtils: typeof StackUtilsType = require('stack-utils');
+
+// Stack cleaner that should remove stack entries *except* the ones from the component itself
+const stackUtils = new StackUtils({
+  cwd: config.paths.base(),
+  internals: StackUtils.nodeInternals().concat(/node_modules/, /react[\\/]test/, /anonymous/, /\butils\b/),
+});
+const reactComponentsRegex = /[\\/]react[/\\]src[/\\]components[\\/]/;
+
+/**
+ * Get the file path within the repo for a component.
+ */
+export default function getComponentPath(Component: React.ComponentType, constructorName: string): string {
+  // Prevent framework code from logging errors
+  const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation();
+  const consoleWarnMock = jest.spyOn(console, 'warn').mockImplementation();
+  // Hack to make UIComponent constructor throw an exception
+  (window as any).gettingComponentPath = true;
+
+  let error = '';
+  let result = '';
+
+  try {
+    // Attempt to cause an exception to be thrown from within the component code, so we can get
+    // the path from the stack.
+    if (Component.prototype) {
+      // Class component. Try a couple different approaches.
+      new (Component as React.ComponentClass)(null); // eslint-disable-line
+      mount(<Component />);
+    } else {
+      // Function component.
+      // Calling as a function (not rendering) will likely result in an invalid hook error.
+      (Component as Function)(null);
+    }
+  } catch (err) {
+    const cleanedStack = stackUtils.clean(err.stack).split(/\r?\n/g);
+    const componentLine = cleanedStack.find(line => reactComponentsRegex.test(line));
+    // const componentLine = stackUtils.clean(err.stack.split(`${err.name}: ${err.message}`)[1].trim()).split(/\r?\n/g)[0];
+    if (reactComponentsRegex.test(componentLine)) {
+      result = stackUtils.parseLine(componentLine).file;
+    } else {
+      error = `getComponentPath algorithm did not return a correct path for ${constructorName}!
+
+Got:
+${componentLine}
+
+From original stack:
+${err.stack}`;
+    }
+  } finally {
+    consoleErrorMock.mockRestore();
+    consoleWarnMock.mockRestore();
+    (window as any).gettingComponentPath = false;
+  }
+
+  if (!result) {
+    // eslint-disable-next-line no-console
+    console.error(error || `getComponentPath algorithm could not find a path for ${constructorName}!`);
+  }
+
+  return result;
+}
```

Comment:
Is there any other way to do it?

Reply:
I haven't found one... :(

Reply:
To be clear I'd very much prefer something less terrible than this, I just couldn't think of anything that didn't involve updating all tests to manually pass the component's path. (But we could do that too.)

Reply:
What about Babel plugin? For tests we can do a special Babel plugin that does this.

The simplified thing that adds `export const __FILEPATH = 'REAL_FILEPATH'`:
```js
const filepathPlugin = ({ types: t }) => ({
  visitor: {
    Program: (path, state) => {
      const declaration = t.exportNamedDeclaration(
        t.variableDeclaration('const', [
          t.variableDeclarator(t.identifier('__FILEPATH__'), t.stringLiteral(state.file.opts.filename)),
        ]),
      );

      path.node.body.push(declaration);
    },
  },
});

module.exports = filepathPlugin;
```

In the similar way we can add static field to React components

Reply:
Interesting idea but I'm hesitant to add a required dep on a babel plugin to conformance tests, since we'd like to make these tests or similar usable for v7 and vnext as well, and at least in the immediate future those won't use babel. (I think also something that makes me nervous about highly customized babel usage in general is that it feels like adding "magic" to the build process. "Magic" features are nice until something breaks and you have to debug, and at that point you'd have to be intimately familiar with the repo-specific build process to understand what's going on.)

Though thinking about it more, @levithomason pointed out in a conversation last week that really most of these tests ought to be converted to lint rules, and then getting the file path wouldn't be an issue. So it might not matter so much which method we choose since it will probably only be temporary.

Reply:
Okay, let's go with the current appoach